### PR TITLE
Replace many usages of Boost with std counterparts

### PR DESCRIPTION
**Pull request by YakoYakoYokuYoku**
Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. Additionally, make sure you've done all of these things:

- [x] I've followed the [contributing guidelines](https://github.com/NatronGitHub/Natron/blob/RB-2.4/CODE_OF_CONDUCT.md) to the best of my understanding
- [x] I've read and understood the [contributing guidelines](https://github.com/NatronGitHub/Natron/blob/RB-2.4/CONTRIBUTING.md)
- [x] I've formatted my code according to Natron's [code style]([#](https://github.com/NatronGitHub/Natron#logistics))
- [x] I've searched the [pull requests tracker](https://github.com/NatronGitHub/Natron/pulls?q=is%3Apr) to ensure that this PR is not a duplicate

## PR Description

**What type of PR is this? (Check one of the boxes below)**

- [x] Breaking change (fix or feature that would cause existing functionality to change)

**What does this pull request do?**

Prior to this PR Natron made usage of many features from the Boost libraries that were not present in the standard C++ library, things like smart pointers or bound functions for example. And now many of them were added to the standard library.

This PR replaces many of the used features of the Boost libraries with modern standard library and bundled utilities. Namely

- Smart pointers (e.g. `boost::scoped_ptr` with `std::unique_ptr`)
- String operations
- Math functions (e.g. `boost::math::cbrt` with `std::cbrt`)
- Bound functions
- Noncopyable subclassing
- C limits (`std::numeric_limits` templates are used instead)

Although Boost still remains in use due to the lack of an equivalent in the stdlib for bidirectional maps and the serialization library is still required because projects use its XML serialization, though both of these things might change in the future.

**Show a few screenshots (if this is a visual change)**

N/A.

**Have you tested your changes (if applicable)? If so, how?**

By building and running both Natron and NatronRenderer.

**Futher details of this pull request**

This PR was split in commits to ease the review process and can be separated into different PRs if requested.

### CODE REVIEW

**Comment by devernay** [comment]
Can you please make the PR on RB-2.4 to avoid code divergence (there may be another 2.4 release)

**Comment by devernay** [comment]
The reason is that we don't have a working RB-2.5 windows build yet

File: CMakeLists.txt [hunk 1]
```diff
@@ -26,7 +26,7 @@ project(Natron
     LANGUAGES CXX C
 )
 
-set(CMAKE_CXX_STANDARD 11)
```
Comment (devernay) [inline CMakeLists.txt:29]: That's a big change, which may be unnecessary.
For which feature exactly is C++14 required?

File: Engine/AbortableRenderInfo.h [hunk 2]
```diff
@@ -30,11 +30,6 @@
 
 #include "Global/GlobalDefines.h"
 
-#if !defined(Q_MOC_RUN) && !defined(SBK_RUN)
```
Comment (devernay) [inline Engine/AbortableRenderInfo.h:33]: are you certain shiboken 1 + shiboken 2 will continue working?
(we're still Qt4 until 2.6)

**Comment by rodlie** [comment]
> The reason is that we don't have a working RB-2.5 windows build yet

I was very close last time I worked on it, but due to various reasons I was not able to finish it. I got very limited time this weekend, but I might have time to finish it on Sunday (no guarantee of course).

**Review by devernay — COMMENTED** [review]
That's a nice code cleanup!

My suggestions/questions:
- why C++14? Which feature is it required for? Remember we are still running releases with Qt4, so C++14 may be a problem.
- apply this to RB-2.4 so that the two codes don't diverge (RB-2.4 is still active because of windows)
- make sure shiboken1 and shiboken2 don't choke (2.4 and 2.5 are still Qt4, sadly, until Qt5 is fully tested and bugs are fixed)

May be part of a future PR:
- add "using namespace std;" at the beginning of cpp files (not headers of course) to improve readability, and remove std:: everywhere

File: CMakeLists.txt (hunk 1 above)
Comment (YakoYakoYokuYoku) [reply CMakeLists.txt:29]: Some of the C++ algorithm library functionality needs the standard to be 14, but from what I see from local tests and GitHub actions C++14 is compatible with Qt4, so there are no problems as of now.

File: Engine/AbortableRenderInfo.h (hunk 2 above)
Comment (YakoYakoYokuYoku) [reply Engine/AbortableRenderInfo.h:33]: Both versions still generate bindings as expected. So I'm certain about that.

**Comment by YakoYakoYokuYoku** [comment]
> Can you please make the PR on RB-2.4 to avoid code divergence (there may be another 2.4 release)

If it's going to be one last release of RB-2.4 then I don't think it _would be worth_ to backport.

File: CMakeLists.txt (hunk 1 above)
Comment (devernay) [reply CMakeLists.txt:29]: you're right, and actually I'm compiling the plugins in C++14 mode (openimageio requires C++14), and I'm compiling Natron on macOS with C++14 too

**Comment by devernay** [comment]
I should be a no-brainer to backport. The codes haven't diverged that much so far

**Comment by YakoYakoYokuYoku** [comment]
Merging in one day if no further objections, if another 2.4 patch release is _really_ needed then I can open another PR, but in the case we have a 2.5 build for Windows then I'll not.

**Review by devernay — APPROVED** [review]

**Pull request merged**